Flash Attention not building

i get always same build error on windows when adding `flash-attn` feature

``` bash
>nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Wed_Oct_30_01:18:48_Pacific_Daylight_Time_2024
Cuda compilation tools, release 12.6, V12.6.85
Build cuda_12.6.r12.6/compiler.35059454_0
```

``` bash
>rustup -V
rustup 1.27.1 (54dd3d00f 2024-04-24)
```
```

  C:\Users\avlin\.cargo\git\checkouts\candle-c6a149c3b35a488f\823a83a\candle-flash-attn\kernels\softmax.h(75): error: identifier "M_LOG2E" is undefined
            const float max_scaled = max(mi) == -((float)(1e+300 * 1e+300)) ? 0.f : max(mi) * (Scale_max ? scale : float(M_LOG2E));
```

# RCA
> For anyone who encourter the same problem

## Root cause
the problem is raised in windows, constant `M_LOG2E` is not defined, because `msvc` disallow VLA, and it works on linux, because on linux `nvcc` uses the `g++` host compiler, and it allows (in a non-standard-compliant way) the use of a VLA in C++ host code.

## Solution
Locate  `softmax.h` file under the `candle-flash-attn` you depend on, add code below
``` c
#ifndef M_LOG2E
#define M_LOG2E 1.4426950408889634074
#endif
```
then remove caches and rebuild it.

---
It might not be the only cross-compile issue you will encounter.
I depend on feature `candle-layer-norm` also, and `nvcc` fail to compile due to  `fatal error C1128: number of sections exceeded object file format limit : compile with /bigobj`


Locate `build.rs` under `candle-layer-norm`, add code below
``` diff
    if should_compile {
        cu_files
            .par_iter()
            .map(|(cu_file, obj_file)| {
                let mut command = std::process::Command::new("nvcc");
                command
                    .arg("-std=c++17")
                    .arg("-O3")
                    .arg("-U__CUDA_NO_HALF_OPERATORS__")
                    .arg("-U__CUDA_NO_HALF_CONVERSIONS__")
                    .arg("-U__CUDA_NO_BFLOAT16_OPERATORS__")
                    .arg("-U__CUDA_NO_BFLOAT16_CONVERSIONS__")
                    .arg("-U__CUDA_NO_BFLOAT162_OPERATORS__")
                    .arg("-U__CUDA_NO_BFLOAT162_CONVERSIONS__")
+                   .arg(format!("--gpu-architecture=sm_{compute_cap}"))
                    .arg("-c")
                    .arg("--compiler-options /bigobj")
                    .args(["-o", obj_file.to_str().unwrap()])
                    .args(["--default-stream", "per-thread"])
                    .arg("--expt-relaxed-constexpr")
                    .arg("--expt-extended-lambda")
                    .arg("--use_fast_math")
                    .arg("--verbose");
```

still compiling...

---
> reference 
> https://github.com/gsgen3d/gsgen/issues/3#issuecomment-1741591484
> https://stackoverflow.com/a/56974843
> https://blog.csdn.net/yyywxk/article/details/140422758
>

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -24,8 +24,8 @@ rust-version = "1.82"
 
 [workspace.dependencies]
 anyhow = "1.0.80"
-candle-core = { git = "https://github.com/EricLBuehler/candle.git", version = "0.8.0", rev = "76819e8" }
-candle-nn = { git = "https://github.com/EricLBuehler/candle.git", version = "0.8.0", rev = "76819e8" }
+candle-core = { git = "https://github.com/EricLBuehler/candle.git", version = "0.8.0", rev = "2002e7a" }
+candle-nn = { git = "https://github.com/EricLBuehler/candle.git", version = "0.8.0", rev = "2002e7a" }
 serde = "1.0.197"
 serde_json = "1.0.114"
 indexmap = { version = "2.2.5", features = ["serde"] }

--- a/mistralrs-core/Cargo.toml
+++ b/mistralrs-core/Cargo.toml
@@ -17,7 +17,7 @@ candle-core.workspace = true
 candle-nn.workspace = true
 serde.workspace = true
 serde_json.workspace = true
-candle-flash-attn = { git = "https://github.com/EricLBuehler/candle.git", version = "0.8.0", rev = "76819e8", optional = true }
+candle-flash-attn = { git = "https://github.com/EricLBuehler/candle.git", version = "0.8.0", rev = "2002e7a", optional = true }
 dirs = "5.0.1"
 hf-hub = { version = "0.3.3", package = "candle-hf-hub" }
 thiserror = "1.0.57"
@@ -81,7 +81,7 @@ llguidance = { git = "https://github.com/microsoft/llguidance", rev = "cfef3df97
 toktrie_hf_tokenizers = { git = "https://github.com/microsoft/llguidance", rev = "cfef3df97372a7b84d74976ff41cc9cb78bca6cc" }
 objc = { version = "0.2.7", optional = true }
 metal = { workspace = true, optional = true }
-candle-flash-attn-v3 = { git = "https://github.com/EricLBuehler/candle.git", version = "0.8.0", rev = "76819e8", optional = true }
+candle-flash-attn-v3 = { git = "https://github.com/EricLBuehler/candle.git", version = "0.8.0", rev = "2002e7a", optional = true }
 safetensors.workspace = true
 serde-big-array = "0.5.1"
 interprocess = "2.2.2"

